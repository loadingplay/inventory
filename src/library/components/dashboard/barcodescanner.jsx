import React from 'react';

export function withBarcodeScanner(WrappedComponent) {
  return class BarCodeScanner extends React.Component {
    constructor(props) {
      super(props);

      this.timer = null;
      this.wrapped = null;
      this.input_string = "";

      this.initListener();
    }

    initListener = () => {
      if (typeof window !== `undefined`)
        document.addEventListener('keyup', this.handleKeyUp);
    }

    handleKeyUp = (e) => {
      const timeoutCallback = () => {
        if (this.input_string.length <= 3) {
          this.input_string = "";
          return;
        }

        this.handleScannerRead(this.input_string);
        this.input_string = "";
      }

      // border cases
      if (e.key === "Enter") timeoutCallback();
      if (this.timer) clearTimeout(this.timer);
      if (e.key.length <= 1) this.input_string += e.key;

      this.timer = setTimeout(timeoutCallback, 100);
    }

    handleScannerRead = (input_string) => {
      if (!this.wrapped || !this.wrapped.handleScannerRead) {
        console.log("implement handleScannerRead method");
        return;
      }

      this.wrapped.handleScannerRead(input_string);
    }

    render = () => {
      return <WrappedComponent {...this.props} ref={e => this.wrapped = e} />;
    }
  }
}
