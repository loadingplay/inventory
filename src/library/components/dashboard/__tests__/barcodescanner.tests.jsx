import React from 'react';
import { mount } from "enzyme";

import { withBarcodeScanner } from "../barcodescanner";

describe("BarcodeScanner", () => {
  const t = withBarcodeScanner(<div />);

  test("it should render properly", () => {
    const wrapper = mount(<t />);
    expect(wrapper).toMatchSnapshot("detault");
  });

  test("it should handle scanner read", () => {
    const wrapper = mount(<t />);
    console.log(wrapper.instance());
    wrapper.instance().handleScannerRead("test");
  });

});
