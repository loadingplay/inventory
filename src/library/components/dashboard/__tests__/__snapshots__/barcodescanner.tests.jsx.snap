// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BarcodeScanner it should render properly: detault 1`] = `<Wrapped />`;
